src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs

```diff
@@ -21,8 +21,8 @@ public sealed class ConfigurationManager : IConfigurationBuilder, IConfiguration
         private readonly ConfigurationSources _sources;
         private readonly ConfigurationBuilderProperties _properties;
 
-        private readonly object _providerLock = new();
-        private readonly List<IConfigurationProvider> _providers = new();
+        // _providerManager provides copy-on-write references. It waits until all readers unreference before disposing any providers.
+        private readonly ProviderManager _providerManager = new();
         private readonly List<IDisposable> _changeTokenRegistrations = new();
```

Comment:
I see `_changeTokenRegistrations` is now being mutated without the lock. Is that going to be a problem?

Reply:
As @stephentoub says, I need to provide better comments about the thread safety guarantees. I'm working on that now.

But to answer the question, the `_changeTokenRegistrations` list is similar to `_sources` in that it's only referenced when config sources are modified and when the ConfigurationManager is disposed. Concurrently modifying sources or disposing while modifying sources was never thread safe. It should still be safe to read config while `_changeTokenRegistrations` is concurrently modified.